File: 0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div2750-ways-to-split-array-into-good-subarrays/0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div2750-ways-to-split-array-into-good-subarrays.cpp
class Solution {
public:
    int dp[100005][2];
    int mod = 1e9+7;
    int n;
    int recur(int idx, int cnt, vector<int> & nums){
        if (idx>=n){
            if (cnt==1){return 1;}
            return 0;
        }

        
        cnt += (nums[idx]==1);
        if (cnt>1){
            return 0;
        }
        
        
        if (dp[idx][cnt]!=-1){
            return dp[idx][cnt];
        }
        int take=0;
        int nottake=0;
        take =recur(idx+1,cnt,nums);
        if (cnt==1){
            nottake = recur(idx+1,0,nums);
        }
        
        return dp[idx][cnt] = (take+nottake)%mod;
        
    }
    
    int numberOfGoodSubarraySplits(vector<int>& nums) {
        n = nums.size();
        memset(dp,-1,sizeof(dp));
        int val = recur(0,0,nums);
        return val%mod;
    }
};

File: 0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div2750-ways-to-split-array-into-good-subarrays/0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div2750-ways-to-split-array-into-good-subarrays.py
class Solution:
    def numberOfGoodSubarraySplits(self, nums: List[int]) -> int:
        n =len(nums)
        MOD = int(1e9)+7
        dp =[[-1]*2 for _ in range(n)]
        def recur(idx,cnt):
            if idx==n:
                if cnt==1:
                    return 1
                return 0
            
            cnt += (nums[idx]==1)
            
            if cnt>1:
                return 0
            
            if dp[idx][cnt]!=-1:
                return dp[idx][cnt]
            
            
            take = recur(idx+1,cnt)
            nottake=0
            if cnt==1:
                nottake = recur(idx+1,0)
            
            dp[idx][cnt] =  (take+nottake)%MOD
            return dp[idx][cnt]
        
        val = recur(0,0)
        return val